### ISSUE TITLE
regression - azurerm_lb - adding frontend_ip_configurations to an existing LB forces replacement

### ISSUE THREAD

**Issue by faust64** [hashicorp/terraform-provider-azurerm#17818]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.0.7

### AzureRM Provider Version

3.16

### Affected Resource(s)/Data Source(s)

azurerm_lb

### Terraform Configuration Files

```hcl
variable "frontends" {
  type = map(
    object({
      # ...
      subnet_id = string
    })
  )
}

resource "azurerm_lb" "loadbalancer" {
  name                = "my-lb"
  location            = "northeurope"
  resource_group_name = "some-resourcegroup"
  sku                 = "Standard"

  dynamic "frontend_ip_configuration" {
    for_each = var.frontends
    content {
      name = frontend_ip_configuration.key
      subnet_id = frontend_ip_configuration.value.subnet_id
      private_ip_address_allocation = "Dynamic"
      public_ip_address_id = null
      zones = [ "1", "2", "3" ]
    }
  }
}
```

### Debug Output/Panic Output

```shell
# Adding a new frontend_ip_configuration

  # module.xxx.azurerm_lb.loadbalancer[0] must be replaced
-/+ resource "azurerm_lb" "loadbalancer" {
...
      + frontend_ip_configuration {
          + gateway_load_balancer_frontend_ip_configuration_id = (known after apply)
          + id                                                 = (known after apply)
          + inbound_nat_rules                                  = (known after apply)
          + load_balancer_rules                                = (known after apply)
          + name                                               = "prod"
          + outbound_rules                                     = (known after apply)
          + private_ip_address                                 = (known after apply)
          + private_ip_address_allocation                      = "dynamic"
          + private_ip_address_version                         = (known after apply)
          + public_ip_address_id                               = (known after apply)
          + public_ip_prefix_id                                = (known after apply)
          + subnet_id                                          = "/subscriptions/xxx/resourceGroups/some-resourcegroup/providers/Microsoft.Network/virtualNetworks/my-vnet/subnets/prod"
          + zones                                              = [
              + "1",
              + "2",
              + "3",
            ] # forces replacement
        }
    }

# Removing an existing frontend_ip_configuration
  # module.xxx.azurerm_lb.loadbalancer[0] must be replaced
-/+ resource "azurerm_lb" "loadbalancer" {
...
      - frontend_ip_configuration {
          - id                            = "yyy" -> null
          - inbound_nat_rules             = [] -> null
          - load_balancer_rules           = [] -> null
          - name                          = "prod" -> null
          - outbound_rules                = [] -> null
          - private_ip_address            = "100.89.4.71" -> null
          - private_ip_address_allocation = "Dynamic" -> null
          - private_ip_address_version    = "IPv4" -> null
          - subnet_id                     = "/xxx" -> null
          - zones                         = [
              - "1",
              - "2",
              - "3",
            ] -> null # forces replacement
        }
    }
```

### Expected Behaviour

Adding or removing frontends from an existing LoadBalancer, I would expect Terraform to preserve the existing loadbalancer intact. Only add or remove frontends.
This used to work with azurerm-2.99, which I'm still using in prod.

### Actual Behaviour

As of upgrading my dev to azurerm-3, I am no longer able to add or remove frontend_ip_configurations in an existing azurerm_lb, without replacing that loadbalancer, reconfiguring all frontends, nlb rules, ...
Swapping from 3 to 4 frontends would involve a 7-8 minutes outage, during which I can't access any of my existing application, LB is gone, or reconfiguring, for quite a long time. While 3-4 frontends is the best case scenario I would have: in prod, some of my LB would have tens of frontends.

As it seems the "zones" argument is the one triggering re-creation of that LB: I also tried to comment it out from my code.
Obviously, changing zones on existing frontend would trigger re-creation as well: so I would add a lifecycle rule ignoring changes... Doing so, plan would work as expected, although apply will fail, as Azure API refuses mixing up zones configuration on a single LB:

```
Error: creating/updating Load Balancer: (Name "my-lb" / Resource Group "xxx"): network.LoadBalancersClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="ResourceAvailabilityZonesCannotBeModified" Message="Resource /subscriptions/xxx/resourceGroups/xxx/providers/Microsoft.Network/loadBalancers/my-lb/frontendIPConfigurations/pod has an existing availability zone constraint 1, 2, 3 and the request has availability zone constraint , which do not match. Zones cannot be added/updated/removed once the resource is created. The resource cannot be updated from regional to zonal or vice-versa." Details=[]
 
  with module.xxx.azurerm_lb.loadbalancer,
  on .terraform/modules/xxx.tf line 2, in resource "azurerm_lb" "loadbalancer":
   2: resource "azurerm_lb" "loadbalancer" {
```

### Steps to Reproduce

plan/apply creating a LB, with frontend input including at least one entry, eg `{ "prod": { subnet_id: <valid-subnet-id> } }`.
then: change input including a second entry. plan output mentions azurerm_lb would be replaced.
Having applied that new frontend: change your input removing an entry. plan output mentions azurerm_lb would be replaced.

proceed with the same steps using azurerm-2.x provider. both adding or removing frontends: azurerm_lb would be updated in-place.

note that both azurerm-2.x and 3.x may show some "weird" diff, suggesting frontend configurations are being swapped (see initial case: https://github.com/hashicorp/terraform-provider-azurerm/issues/17809 ), although those diffs are misleading: running apply, terraform only adds the new configuration, there's no change to existing relays.

### References

Re-formatting https://github.com/hashicorp/terraform-provider-azurerm/issues/17809. Initially filed for some weird diff with 2.99, although troubleshooting and testing azurerm-3.x: found a regression.

**Comment by faust64** [comment]
any news about this?
I wouldn't want bot to auto-close this for inactivity ...

This is still very much an issue to us.
I can't tell my ops that, "as of upgrading azurerm provider" they can no longer reconfigure loadbalancers, on production platforms (we have 100s of LBs), while we had no issues operating those with azurerm-2.x.

**Comment by faust64** [comment]
running azurerm 3.24, still an issue.
any news?

**Comment by tombuildsstuff** [comment]
hi @faust64 

Zones can only be specified for a Frontend IP Configuration when they're specified for the LoadBalancer itself, since these are defined for the Frontend IP Configuration but aren't for the Load Balancer itself, these aren't set by the Azure API and highlight as a diff here. Since Zones can only be changed by recreating the resource, the Terraform Plan is correctly highlighting that this needs to be recreated to set these fields (although unfortunately we can't highlight these both of these fields must be set together).

[More information on the changes to Zones can be found in the 3.0 upgrade guide](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/3.0-upgrade-guide#behavioural-updates) - however this can be fixed by either removing `zones` from the Frontend IP Configuration block, or by specifying them against the Load Balancer (and recreating it) such that the Load Balancer and Frontend IP Configuration both support zones. Since this is a question regarding Terraform Configuration rather than a bug in the Azure Provider, I'm going to close this issue for the moment - but should you have further questions I'd direct you towards [the Community Discuss Forum](https://discuss.hashicorp.com).

Thanks!

**Comment by faust64** [comment]
> this can be fixed by either removing zones from the Frontend IP Configuration block

You kidding, right?

upgrading from azurerm-2 to azurerm-3.
either I leave my code as it is, never setting those zones in frontend ip configuration. Then terraform wants to re-create half of  my existing LB
Or I put those zones, in each frontend, aligning with what terraform tells me "changed outside" of its state, to avoid my LB being re-created. And still, terraform wants to re-create another half of my LB.
So... upgrading, we came to the conclusion: add those zones in all frontend_ip_configuration blocks. AND, add ignore_lifecycle_change, ignoring any changes to existing frontend_ip_configuration.zones.

"removing the zones", as you suggest, is just not realistic. Actually, it's even worse: I would need to re-create all my loadbalancers. This is just out of the question. The reason I added those stupid zones everywhere, today, is precisiely because of upgrading from 2.x to 3.x.

While first running terraform plan with azurerm-3, those zones where everywhere in terraform diffs. "has changed outside of terraform state". And "will change from [] to [1,2,3]", or [1,2,3] to []", "triggers force-replace". Also: depending on how old my LBs are, terraform would tell me that some frontend_ip_configuration has zones 1,2,3, while another has zones []. This, with the exact same input passed in creating those LB. Your provider is not consistent. Your code is bugged. This has nothing to do with my configuration.

We've been ignoring lifecycle_changes to any of those zones.
The only place we did set those zones, is on recently-introduced frontend ip configurations (maybe 10 in the last months) ... while I have litteraly 1000s frontend ip configurations, provisioned with azurerm-v2, for which the only way I had to go to v3, without re-creating them, was to ignore lifecycle changes. Thus: good news, zones ain't there/at least not configured by our code, no worries. Still, my issue very much is present. What should I do, then?

**Comment by faust64** [comment]
still no news
still a regression
that was introducing upgrading from azurerm-2, to azurerm-3
I am aware of https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/3.0-upgrade-guide#behavioural-updates
Which is just wrong, when using dynamic frontends

**Comment by tombuildsstuff** [comment]
@faust64 

Whilst I'm sorry that you're frustrated with this, unfortunately this behaviour ultimately comes from the Azure API - where the behaviour of the API changes depending on when the Load Balancer was created (or rather, which API version was used to create it, and when - with some versions requiring no zones, some requiring a single zone, others requiring multiple zones).

The Azure API itself now requires that when the Load Balancer itself has Zones that all items nested under it (such as Frontend IP Configurations) must also have Zones. Unfortunately the behaviour of the API changes depending on which version of the API was used to provision the Load Balancer, such that the Azure API can opt to hide the implied Zones from nested resources until there's a change to them in older versions - but return/require them for newer versions.

Version 2.x of the Azure Provider opted to mask the behaviour of Zones somewhat by having this both Optional and Computed, meaning that if the value wasn't defined then we wouldn't remove it - the challenge with this approach is that the API can default to using Zones in some Regions but not in others, leading to confusion. We also tried to workaround the behaviour of the Azure API here and omit the diff as required (trying to specify Zones where required and not where they weren't) which ended up causing more confusion and was removed in v3.0.

Version 3.0 of the Azure Provider by contrast intentionally removes Computed from these so that we highlight that there's a diff if these aren't specified.

Whilst neither of these approaches is ideal, ultimately the Azure API is the root-cause of this behaviour - and unfortunately this isn't something that we can workaround in the Provider (aside from making this field non-computed to highlight that it has to be set, which we've done in v3.0).

From Terraform's side this means that the diff can be removed by ensuring that the `zones` argument is specified to match the current state of the Resource - which is the recommended approach here and is mentioned in the upgrade guide.

As mentioned above, whilst I can appreciate this is frustrating I believe this can be fixed by specifying `zones` as appropriate  in the Terraform Configuration. If that doesn't fix this I'd recommend reaching out to Azure Support, since otherwise I believe this Load Balancer would need to be recreated (using the latest version of the Provider/API) to obtain the latest behaviour from the Azure API.

Thanks!
